Mostly read-only REST API

bonzini opened the pull request:
This pull request implements the beginning of #25 ("Add REST API").
The API is minimal on the write side (no login, no posting of messages), but it already includes:
* the possibility to search and browse the database
* a complex schema to easily retrieve the metadata for entire threads (aka "series") with a limited number of API calls
* plugin integration
* an embryo of a first-class "results" model, used by the git and testing plugins to hide the internal properties-based implementation

bonzini commented:
The failure is only in Python 3.7 nightly builds, and is due to Django, not this pull request:

      File "/home/travis/virtualenv/python3.7-dev/lib/python3.7/site-packages/django/contrib/admin/widgets.py", line 152
        '%s=%s' % (k, v) for k, v in params.items(),
        ^
    SyntaxError: Generator expression must be parenthesized

famz commented on api/rest.py:
```diff
@@ -0,0 +1,60 @@
+
+class IsMaintainerOrReadOnly(permissions.BasePermission):
+    """
+    Allows access only to admin users.
```
"only to maintainers"?

bonzini replied:
"only to admin users or maintainers"

famz commented on api/rest.py:
```diff
@@ -127,11 +153,26 @@ class SeriesSerializer(BaseMessageSerializer):
+        return results
+    results = SerializerMethodField()
+
+    def __init__(self, *args, **kwargs):
```
Is it easier to read if we move __init__ to the first method in the class?

famz commented:
I haven't looked in depth into django-rest-framework and the nested router plugin. But the PR looks great in general.

The question I have is testing. Is the plan to swtich patchew-cli or the web interface to be a client? (I think both are good things) And are you going to add cases to tests/?

bonzini commented:
> Is the plan to swtich patchew-cli or the web interface to be a client? (I  think both are good things)

Yes and no respectively (I prefer the web interface to do everything server-side, I don't like single-page web apps and they are beyond my limited knowledge of Django!).

However, even for patchew-cli, people are using patchew-project/patchew's patchew-cli as a canonical source of "a command line tha works with patchew.org" so that should only be done after the REST API is deployed on patchew.org.

> And are you going to add cases to tests/?

Yes, and I'll do that before pushing this.

bonzini commented:
Needless to say, adding tests found some bugs. :) Comparison between v1 and v2:

https://github.com/bonzini/patchew/compare/rest-v1...bonzini:rest?expand=1

famz commented on tests/test_rest.py:
```diff
@@ -0,0 +1,102 @@
+# Copyright 2017 Red Hat, Inc.
+#
+# Authors:
+#     Fam Zheng <famz@redhat.com>
```
I don't mind but you can update the year and the name if you want. :)

famz commented on tests/test_git.py:
```diff
@@ -100,6 +102,51 @@ def test_apply_with_base_and_brackets(self):
+If you prefer to skip this patch, run "git am --skip" instead.
+To restore the original branch and stop patching, run "git am --abort".
+Failed to apply patch:
+[Qemu-devel] [PATCH] bar patch""")
```
I think the actual error message depends on the git version. I'm getting this on Fedora (git version 2.14.3):

```
Switched to a new branch '20160628014747.20971-2-famz@redhat.com'
Branch 20160628014747.20971-2-famz@redhat.com set up to track local branch test by rebasing.
Applying: bar patch
error: sha1 information is lacking or useless (foo).
error: could not build fake ancestor
Patch failed at 0001 bar patch
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
Failed to apply patch:
[Qemu-devel] [PATCH] bar patch
```

I.e. there is one extra line in the middle:

```
Branch 20160628014747.20971-2-famz@redhat.com set up to track local branch test by rebasing.
```
Maybe simply remove the `self.assertEquals(log.content, expected_log)` check?

bonzini replied:
I'll change it to a status code test (must be 200).

famz commented on tests/test_git.py:
```diff
@@ -100,6 +102,51 @@ def test_apply_with_base_and_brackets(self):
+To XXXXXX/repo
+ * [new tag]         patchew/20160628014747.20971-1-famz@redhat.com -> patchew/20160628014747.20971-1-famz@redhat.com
+""")
+        self.assertEquals(log.content, expected_log)
```
The same error here for me.

famz commented:
I am done with an incremental review, the code looks good. Thank you for working on this, and adding tests! I think we are almost ready to go!

bonzini commented:
Done! As before, for an incremental review you can use branch `rest-v2` (I rebased so that travis passes, but the changes to the series are identified easily).

As discussed on IRC, I've added a migration to fix UTF-8 recipients and a test case for that too. You still cannot search for UTF-8 recipients because of how they are encoded in JSON works, but that's a separate bug in `api.search`.

famz commented:
Looks good. Except this error when migrating the live patchew.org database (let me know if you want access the server, or alternatively):

```
(venv) fam@lemon:~/work/patchew$ ./manage.py migrate
Loaded module: footer
Loaded module: diff
Loaded module: email
Loaded module: git
Loaded module: tags
Loaded module: testing
Operations to perform:
  Apply all migrations: admin, api, auth, contenttypes, sessions
Running migrations:
  Applying api.0020_auto_20180204_0647... OK
  Applying api.0021_fix_recipients_utf8...Traceback (most recent call last):
  File "./manage.py", line 19, in <module>
    execute_from_command_line(sys.argv)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/migration.py", line 129, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/stor/work/patchew/venv/lib64/python3.6/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/stor/work/patchew/api/migrations/0021_fix_recipients_utf8.py", line 24, in fix_utf8_recipients
    recipients = [[_parse_header(x[0]), x[1]] for x in recipients]
  File "/stor/work/patchew/api/migrations/0021_fix_recipients_utf8.py", line 24, in <listcomp>
    recipients = [[_parse_header(x[0]), x[1]] for x in recipients]
  File "/stor/work/patchew/api/migrations/0021_fix_recipients_utf8.py", line 12, in _parse_header
    h = h.decode(c or 'utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 8: invalid continuation byte
```

bonzini commented:
I'll drop the last patch then. I think the culprit is [this patch](http://patchew.org/QEMU/20170615220501.GA26408@flamenco/mbox).

bonzini commented:
Oh, it's actually a wrong record in the database:

                    "name": "=?UTF-8?q?Daniel=20D=EDaz?=",
                    "address": "daniel.diaz@linaro.org"

Thunderbird "correctly" shows it as _Daniel D�az_.
